modules/flatgeobuf/src/lib/binary-geometries.js

```diff
@@ -0,0 +1,162 @@
+  // contains the end indices of each ring of the Polygon. geometry.endsArray()
+  // omits the initial 0 that we have in our internal format.
+  const ends = (geometry.endsArray() && Array.from(geometry.endsArray())) || [xy.length / 2];
+  ends.unshift(0);
+
+  const primitivePolygonIndices = {value: new Uint16Array(ends), size: 1};
+  const polygonIndices = {value: new Uint16Array([0, xy.length / 2]), size: 1};
+
+  return {
+    positions,
+    primitivePolygonIndices,
+    polygonIndices
+  };
+}
+
+// eslint-disable-next-line max-statements
+function parseMultiPolygons(geometry) {
+  // Create arrays for each geometry part, then concatenate
+  const parsedParts = [];
+  let nPositions = 0;
+  let nPrimitivePolygonIndices = 1;
+  let nPolygonIndices = 1;
+
+  for (let i = 0; i < geometry.partsLength(); i++) {
+    const part = geometry.parts(i);
+    const polygon = parsePolygons(part);
+
+    nPositions += polygon.positions.value.length;
+    nPrimitivePolygonIndices += polygon.primitivePolygonIndices.value.length - 1;
+    nPolygonIndices += polygon.polygonIndices.value.length - 1;
+
+    parsedParts.push(polygon);
+  }
+
+  const concatPositions = new Float64Array(nPositions);
+  const concatPrimitivePolygonIndices = new Uint32Array(nPrimitivePolygonIndices);
+  const concatPolygonIndices = new Uint32Array(nPolygonIndices);
+
+  let positionCounter = 0;
+  let primitivePolygonIndicesCounter = 1;
+  let polygonIndicesCounter = 1;
+
+  // Assumes all parts of the multipolygon have the same size
+  const positionSize = parsedParts[0].positions.size;
+
+  for (const parsedPart of parsedParts) {
+    concatPositions.set(parsedPart.positions.value, positionCounter * positionSize);
+
+    // For indices, need to add positionCounter so that position indices are
+    // correct in the concatenated positions
+    concatPrimitivePolygonIndices.set(
+      parsedPart.primitivePolygonIndices.value.subarray(1).map(x => x + positionCounter),
+      primitivePolygonIndicesCounter
+    );
+    concatPolygonIndices.set(
+      parsedPart.polygonIndices.value.subarray(1).map(x => x + positionCounter),
+      polygonIndicesCounter
+    );
+
+    positionCounter += parsedPart.positions.value.length / positionSize;
+    primitivePolygonIndicesCounter += parsedPart.primitivePolygonIndices.value.length - 1;
+    polygonIndicesCounter += parsedPart.polygonIndices.value.length - 1;
+  }
+
+  return {
+    positions: {value: concatPositions, size: positionSize},
+    primitivePolygonIndices: {value: concatPrimitivePolygonIndices, size: 1},
+    polygonIndices: {value: concatPolygonIndices, size: 1}
+  };
+}
+
+// Combine xy and z arrays
+function blitArrays(xy, z) {
+  if (!z) {
+    return {value: xy, size: 2};
+  }
+
+  if (z.length * 2 !== xy.length) {
+    throw new Error("Z array must be half XY array's length");
+  }
+  const totalLength = xy.length + z.length;
+
+  const xyz = new Float64Array(totalLength);
```

Comment:
Generating a typed array for every little feature is unlikely to gain us any performance advantage.

Is a 2 pass approach out of the question?

Reply:
The Flatgeobuf module parses features individually. For non-streaming you could theoretically call `deserialize` twice, once to collect lengths, again to set positions within an instantiated typed array.

I'm not sure how you could do that with streaming. Flatgeobuf's `deserializeStream` abstracts this from the user. The parsing function you pass operates on a single feature at a time.

Note that for `positions` of all 2D geometry types except for MultiPolygon, no copy should be needed. They're already transported in a flat 2D array.